### PR TITLE
[inductor] Add realize_hint and small fixes

### PR DESCRIPTION
**Pull request by jansel**

### CODE REVIEW

File: torchinductor/codegen/triton.py [hunk 1]
```diff
@@ -510,7 +510,9 @@ def getter(flat_vars):
 
             return_getters_groups.append(return_getters)
 
-        assert all(s == 1 for s in remaining)
+        assert all(
+            V.graph.sizevars.size_hint(s) == 1 for s in remaining
+        ), f"failed to set ranges {remaining} {lengths}"
```
Comment (desertfire) [inline torchinductor/codegen/triton.py:515]: This is already in the tip. Is this the main fix for https://github.com/pytorch/torchdynamo/issues/661, or is `realize_hint` the main fix?

File: torchinductor/ir.py [hunk 2]
```diff
@@ -2653,27 +2656,43 @@ def realize(self):
         self.data.name = V.graph.register_buffer(self.data)
         return self.data.name
 
+    def realize_hint(self):
+        """
+        Called on buffers we expect to be forced to realize later.
+        """
+        if self.num_reads() > 1:
```
Comment (desertfire) [inline torchinductor/ir.py:2663]: I am trying to relate this to some recent discussion in our group chat. Does this change mean we are not leveraging DSE anymore? Is there any performance implication here?

Comment (desertfire) [reply torchinductor/ir.py:2663]: Oh, I see you are replacing certain `realize` with `realize_hint`, so this should only make perf better.

File: torchinductor/ir.py [hunk 3]
```diff
@@ -2653,27 +2656,43 @@ def realize(self):
...
+                num_reads > config.realize_reads_threshold
                 or len(self.inner_fn_str()) > config.realize_bytes_threshold
             ):
                 self.realize()
 
+    def num_reads(self):
+        if isinstance(
+            self.data, (ComputedBuffer, InputsKernel, InputBuffer, ReinterpretView)
+        ):
+            return 1
```
Comment (desertfire) [inline torchinductor/ir.py:2683]: Why return 1 for `ComputedBuffer`? Should we recursively call `num_reads` on it?

Comment (jansel) [reply torchinductor/ir.py:2691]: Fixed

**Review by desertfire — APPROVED** [review]

**Pull request merged**